### ISSUE TITLE
Allow expressions to compute keys

### ISSUE THREAD

**Issue by larsga** [schibsted/jslt#56]
Issue #55 demonstrates that there might be a need for being able to write objects like this:

```
{
  .foo : .bar,
  "something" : .else
}
```

There doesn't seem to be any clear reason why this couldn't be supported.

**Comment by ngsoftwaredev** [comment]
That feature would help me so much!

**Comment by larsga** [comment]
Could you show an example of how you would use it? That would be helpful for me.

**Comment by ngsoftwaredev** [comment]
One use case I have is to generate Elasticsearch queries from a business-specific DSL, using JSLT, which allows me for far more sophisticated templating than Elasticsearch native templating (based on mustache.js)!

Here's a sample JSLT :

```
let searchFields = [ "title", "description", "category" ]

let searchText = .text

{
  "query": {
    "dis_max": {
      "tie_breaker": 1,
      "queries": [ for ($searchFields)
        {
          "match": {
            . + ".lang"  : {
              "query": $searchText
            }
          }
        }        
      ]
    }
  },
  "highlight": {
    "fields": { for ($searchFields)
      . + ".lang": { "number_of_fragments": 1 }
    }
  }
}
```

This does not compile. This is only avery simple snippet, the query could be much more complex, but I have always the need to have some variable (a field name) as a key to specifiy a search operation on a given field.

Currently I have to build a JSON string instead and the use the `from-json` function.

**Comment by larsga** [comment]
This makes sense. Thank you! Yes, I think we should add support for this. I'll see if I can get it done.

**Comment by larsga** [comment]
Now I find I need this feature myself.

**Comment by larsga** [comment]
Some issues that need to be worked out.

Dynamic keys in objects that use the object matcher. Since we don't know at compile-time what keys are defined in the matcher the matching behaviour becomes more complex. Possible solution: forbid dynamic keys in matchers.

What if the key expression produces something that's not a string? We consider this an error.

What if the key expression ends up producing a duplicate key? Duplicate literal keys are forbidden at  the moment.  Consider this an error?

**Comment by larsga** [comment]
Fixed in commit 7342c95
